### PR TITLE
add-k8s cmd supports to gather regions/zones

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

add-k8s cmd gets regions/zones from `--regions` options or gather cluster facts via provider(k8s) api.

## QA steps
```
1. prepare a gke/aks k8s cluster;
2. juju add-k8s gke1 --debug --region  gce/<region>;  // this will use the region option value;
3. juju add-k8s aks1 --debug --region  azure/<region>;  // this will use the region option value;
4. juju add-k8s gke1 --debug;  // juju will try to fetch region meta data from the gke k8s api
5. juju add-k8s aks1 --debug;  // juju will try to fetch region meta data from the aks k8s api;
6. juju add-k8s gke1 --debug --region  invalid-cloud-region; // valid format - <cloud>/<region>;
```

## Documentation changes

new option `--region` to let user to parse k8s cluster host cloud type and region;
if it's not specified, juju will try to ask k8s cluster for it;

## Bug reference

new feature.

### CODE REVIEW

**Comment by ycliuhw** [comment]
[public k8s login, gke, aks etc](https://trello.com/c/9vm0Ep7E/554-public-k8s-login-gke-aks-etc)

File: caas/kubernetes/clientconfig/k8s.go [hunk 1]
```diff
@@ -102,7 +102,7 @@ func cloudsFromConfig(config *clientcmdapi.Config) (map[string]CloudConfig, erro
 func credentialsFromConfig(config *clientcmdapi.Config) (map[string]cloud.Credential, error) {
 	rv := map[string]cloud.Credential{}
 	for name, user := range config.AuthInfos {
-		logger.Debugf("name %q, user %#v", name, user)
+		logger.Debugf("name %q", name)
```
Comment (wallyworld) [inline caas/kubernetes/clientconfig/k8s.go]: can we retain user info in debug?

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -179,6 +182,74 @@ func (k *kubernetesClient) PrepareForBootstrap(ctx environs.BootstrapContext) er
 	return nil
 }
 
+func newLabelRequirements(rs ...requirement) k8slabels.Selector {
+	s := k8slabels.NewSelector()
+	for _, r := range rs {
+		l, err := k8slabels.NewRequirement(r.key, r.operator, r.strValues)
+		if err != nil {
+			panic(errors.Annotate(err, "fix this"))
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: yikes!

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -179,6 +182,74 @@ func (k *kubernetesClient) PrepareForBootstrap(ctx environs.BootstrapContext) er
...
+	return s
+}
+
+type requirement struct {
+	key       string
+	operator  selection.Operator
+	strValues []string
+}
+
+var checkers = map[string]k8slabels.Selector{
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: let's move all this new stuff to a separate file

File: caas/kubernetes/provider/k8s.go [hunk 4]
```diff
@@ -179,6 +182,74 @@ func (k *kubernetesClient) PrepareForBootstrap(ctx environs.BootstrapContext) er
...
+		}
+	}
+	return ""
+}
+
+var regionLabelName = "failure-domain.beta.kubernetes.io/region"
+
+// ListHostCloudRegions lists all the cloud regions that this cluster has worker nodes/instances running in.
+func (k *kubernetesClient) ListHostCloudRegions() (set.Strings, error) {
+	// we only check 5 nodes which should be sufficient.
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Where did 5 come from?
That doesn't seem sufficient to cover everything?

File: cloud/clouds.go [hunk 5]
```diff
@@ -111,6 +121,9 @@ type Cloud struct {
 	// environs.RegisterProvider.
 	Type string
 
+	// HostCloud represents the k8s host cloud. The format is <cloudName>/<region>.
+	HostCloud string
```
Comment (wallyworld) [inline cloud/clouds.go]: I think this should be called
HostCloudRegion

File: caas/kubernetes/provider/k8s.go (hunk 2 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: moved to init.go to fail it earlier if it' has invalid conf. and bettered error msg.

**Comment by ycliuhw** [comment]
!!build!!

File: cmd/juju/caas/add.go [hunk 6]
```diff
@@ -54,7 +70,8 @@ can contain definitions for different k8s clusters, use --cluster-name to pick
 which one to use.
 
 Examples:
-    juju add-k8s myk8scloud
+	juju add-k8s myk8scloud
+	juju add-k8s myk8scloud --region <cloudType/region>
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: should add some text to the usageAddCAASDetails above to explain why region is needed, ie when running add-k8s on jaas and the cloud/region cannot be detected automatically

File: cmd/juju/caas/add.go [hunk 7]
```diff
@@ -114,6 +143,7 @@ func (c *AddCAASCommand) Info() *cmd.Info {
 func (c *AddCAASCommand) SetFlags(f *gnuflag.FlagSet) {
 	c.CommandBase.SetFlags(f)
 	f.StringVar(&c.clusterName, "cluster-name", "", "Specify the k8s cluster to import")
+	f.StringVar(&c.cloudRegion, "region", "", "kubernetes cluster cloud region")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: ...cloud and region

File: cmd/juju/caas/add.go [hunk 8]
```diff
@@ -195,19 +225,34 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
 		Type:           c.caasType,
 		Endpoint:       currentCloud.Endpoint,
-		AuthTypes:      []cloud.AuthType{credential.AuthType()},
+		AuthTypes:      []jujucloud.AuthType{credential.AuthType()},
 		CACertificates: []string{cloudCAData},
 	}
 
+	cloudRegion := c.cloudRegion
+	if cloudRegion == "" {
+		cloudRegion, err = c.getClusterRegion(ctx, newCloud, credential)
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: We can even store there error here to print out later if required. ie it's done here as debug but if the jaas controller needs it and we couldn't get it, we can print the actual error

File: cmd/juju/caas/add.go [hunk 9]
```diff
@@ -195,19 +225,34 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+		}
+		newCloud.HostCloudRegion = cloudRegion
+	}
+
 	if err := addCloudToLocal(c.cloudMetadataStore, newCloud); err != nil {
 		return errors.Trace(err)
 	}
 
-	cloudClient, err := c.apiFunc()
+	cloudClient, err := c.addCloudAPIFunc()
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: so below we want to introduce a new MissingCloudRegion error - juju will never return it but jaas can - the error can be defined next to the Cloud struct. if we get that error, we can then also see if there ws an error getting the cloud region and also print that, or else just inform the user a cloud region is needed

File: cmd/juju/caas/add.go [hunk 10]
```diff
@@ -228,6 +273,115 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+	defer errors.DeferredAnnotatef(&err, "validating cloud region %q", cloudRegion)
+
+	cloudType, region, err := parseCloudRegion(cloudRegion)
+	if err != nil {
+		return errors.Annotate(err, "parsing cloud region")
+	}
+
+	clouds, err := c.getAllCloudDetails()
+	if err != nil {
+		return errors.Annotate(err, "listing juju cloud regions")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: don't need the word "juju" here

File: cmd/juju/caas/add.go [hunk 11]
```diff
@@ -228,6 +273,115 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+
+	result := make(chan string, 1)
+	errChan := make(chan error, 1)
+	go func() {
+		cloudRegions, err := broker.ListHostCloudRegions()
+		if err != nil {
+			errChan <- err
+		}
+		if cloudRegions == nil || cloudRegions.Size() == 0 {
+			errChan <- errors.New("no cloud region information is set in the cluster")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: i this really an error?

File: cmd/juju/caas/add.go [hunk 12]
```diff
@@ -228,6 +273,115 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+		} else {
+			// we currently assume it's always a single region cluster.
+			result <- cloudRegions.SortedValues()[0]
+		}
+	}()
+
+	timeout := 30 * time.Second
+	defer fmt.Println("")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: we should print to ctx.Stdout etc

File: cmd/juju/caas/add.go [hunk 13]
```diff
@@ -9,40 +9,56 @@ import (
...
+	"time"
 
 	"github.com/juju/cmd"
 	"github.com/juju/errors"
 	"github.com/juju/gnuflag"
 	"github.com/juju/loggo"
+	"github.com/juju/utils/set"
 	"golang.org/x/crypto/ssh/terminal"
-	"gopkg.in/juju/names.v2"
+	names "gopkg.in/juju/names.v2"
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: no need for alias

File: cmd/juju/caas/add.go [hunk 14]
```diff
@@ -75,10 +94,19 @@ type AddCAASCommand struct {
 	// clusterName is the name of the cluster (k8s) or credential to import
 	clusterName string
 
+	// cloudRegion is the cloud region that the nodes of cluster (k8s) are running in.
+	// The format is <cloudType/region>
+	cloudRegion string
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: We use "hostCloudRegion" elsewhere.
Nice to be consistent here too.

File: cmd/juju/caas/add.go [hunk 15]
```diff
@@ -195,25 +227,47 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
-	if err := addCloudToController(cloudClient, newCloud); err != nil {
+	if err := c.addCloudToControllerWithRegion(cloudClient, newCloud); err != nil {
+		if false {
+			// try to fetch cloud and region then retry.
+			// TODO(caas): once jaas controller support errors.IsRegionRequired(err), enable here.
+			cloudRegion, err := c.getClusterRegion(ctx, newCloud, credential)
+			errMsg := `
+Jaas requires cloud and region information. But it's
+not possible to fetch cluster region in this case, 
+	please use --region option to parse in if you want to
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: please use --region to specify the cloud/region manually

File: cmd/juju/caas/add_test.go [hunk 16]
```diff
@@ -15,22 +15,28 @@ import (
 	"github.com/juju/loggo"
 	jujutesting "github.com/juju/testing"
 	jc "github.com/juju/testing/checkers"
+	"github.com/juju/utils/set"
 	gc "gopkg.in/check.v1"
-	"gopkg.in/juju/names.v2"
+	names "gopkg.in/juju/names.v2"
```
Comment (wallyworld) [inline cmd/juju/caas/add_test.go]: no need for alias

File: cmd/juju/caas/add_test.go [hunk 17]
```diff
@@ -15,22 +15,28 @@ import (
...
+	yaml "gopkg.in/yaml.v2"
```
Comment (wallyworld) [inline cmd/juju/caas/add_test.go]: and here

File: cmd/juju/caas/add.go [hunk 18]
```diff
@@ -228,6 +282,127 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+func newK8sBrokerGetter(rootAPIGetter func() (api.Connection, error)) BrokerGetter {
+	return func(cloud jujucloud.Cloud, credential jujucloud.Credential) (k8sBrokerRegionLister, error) {
+		conn, err := rootAPIGetter()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		modelAPI := modelconfig.NewClient(conn)
+		defer modelAPI.Close()
+
+		// this current model is not really we want, but it's just for generating the config.
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: We can reword this comment.
// Use the controller model config for constructing the Juju k8s client.

File: cmd/juju/caas/add.go [hunk 19]
```diff
@@ -228,6 +282,127 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+		if v.CloudType == cloudType {
+			for k := range v.RegionsMap {
+				if k == region {
+					logger.Debugf("cloud region %q is valid", cloudRegion)
+					return nil
+				}
+			}
+		}
+	}
+	return errors.NotValidf("")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: errors.NotValidf("cloud region %s", cloudRegion)

File: cmd/juju/caas/add.go [hunk 20]
```diff
@@ -195,25 +227,47 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+			errMsg := `
+Jaas requires cloud and region information. But it's
+not possible to fetch cluster region in this case, 
+	please use --region option to parse in if you want to
+`[1:]
+			if err != nil {
+				return errors.Annotate(err, errMsg)
+			}
+			if cloudRegion == "" {
+				return errors.NotValidf(errMsg)
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: errors.NewNotValid(nil, errMsg)

File: cmd/juju/caas/add.go [hunk 21]
```diff
@@ -228,6 +282,127 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+			return nil, errors.Trace(err)
+		}
+		return caas.New(environs.OpenParams{Cloud: cloudSpec, Config: cfg})
+	}
+}
+
+func parseCloudRegion(cloudRegion string) (string, string, error) {
+	fields := strings.SplitN(cloudRegion, "/", 2)
+	if len(fields) != 2 || fields[0] == "" || fields[1] == "" {
+		return "", "", errors.NotValidf("expected <cloudType>/<region>")
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: errors.NotValidf("cloud region %s", cloudRegion)

File: cmd/juju/caas/add_test.go [hunk 22]
```diff
@@ -156,6 +172,11 @@ func (fcs *fakeCredentialStore) UpdateCredential(cloudName string, details cloud
 	fcs.AddCall("UpdateCredential", cloudName, details)
 	return nil
 }
+func (s *addCAASSuite) SetUpSuite(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/caas/add_test.go]: This should be in SetUpTest
we want to reset the current regions for each test

File: cmd/juju/caas/add_test.go [hunk 23]
```diff
@@ -280,6 +329,152 @@ func (s *addCAASSuite) TestNonExistClusterName(c *gc.C) {
...
+	regionStr      string
+	expectedErrStr string
+}
+
+func (s *addCAASSuite) TestRegionFlag(c *gc.C) {
+	for _, ts := range []regionTestCase{
+		{
+			title:          "missing cloud",
+			regionStr:      "/region",
+			expectedErrStr: `validating cloud region "/region": parsing cloud region: expected <cloudType>/<region> not valid`,
```
Comment (wallyworld) [inline cmd/juju/caas/add_test.go]: This error message is poorly worded, hence the request t change the code to construct the NotValid error. see how " not valid" is appended? errors.NotValidf("foo") results in "foo not valid". If you want a NotValid error with a raw string, you use errors.NewNotValid(nil, "foo")

File: cmd/juju/caas/add.go [hunk 24]
```diff
@@ -195,25 +228,46 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
 
-	if err := addCloudToController(cloudClient, newCloud); err != nil {
+	if err := c.addCloudToControllerWithRegion(cloudClient, newCloud); err != nil {
+		if params.IsCodeCloudRegionRequired(err) {
+			// try to fetch cloud and region then retry.
+			cloudRegion, err := c.getClusterRegion(ctx, newCloud, credential)
+			errMsg := `
+Jaas requires cloud and region information. But it's
+not possible to fetch cluster region in this case, 
+	please use --region to specify the cloud/region manually
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: Extra whitespace is a mistake?
Should also use a "." at the end

File: cmd/juju/caas/add.go [hunk 25]
```diff
@@ -195,25 +228,46 @@ func (c *AddCAASCommand) Run(ctxt *cmd.Context) error {
...
+				return errors.Annotate(err, errMsg)
+			}
+			if cloudRegion == "" {
+				return errors.NewNotValid(nil, errMsg)
+			}
+			newCloud.HostCloudRegion = cloudRegion
+			if err := c.addCloudToControllerWithRegion(cloudClient, newCloud); err != nil {
+				return errors.Trace(err)
+			}
+		} else {
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: We can reduce the amount of indentation by inverting the if condition
```
  if !params.IsCodeCloudRegionRequired(err) {
    return errors.Trace(err)
  }
  // try to fetch cloud and region then retry.
  ....
```

File: cmd/juju/caas/add_test.go [hunk 26]
```diff
@@ -15,22 +15,30 @@ import (
...
-	fileCredentialStore *fakeCredentialStore
-	fakeK8SConfigFunc   clientconfig.ClientConfigFunc
+	dir                       string
+	fakeCloudAPI              *fakeAddCloudAPI
+	fakeK8sBrokerRegionLister *fakeK8sBrokerRegionLister
+	store                     *fakeCloudMetadataStore
+	fileCredentialStore       *fakeCredentialStore
+	fakeK8SConfigFunc         *clientconfig.ClientConfigFunc
+	currentClusterRegionSet   *set.Strings
+	isCloudRegionRequired     *bool
```
Comment (wallyworld) [inline cmd/juju/caas/add_test.go]: This doesn't need to be a pointer.
Just set it to true when needed

File: cmd/juju/caas/add.go (hunk 24 above)
Comment (ycliuhw) [reply cmd/juju/caas/add.go]: ah, intent to highlight by the indent but happy to remove.

File: cmd/juju/caas/add_test.go (hunk 26 above)
Comment (ycliuhw) [reply cmd/juju/caas/add_test.go]: `isCloudRegionRequired` is passed into `fakeAddCloudAPI` in `SetUpTest`. This pointer allows to  reset value in actual test later after SetUpTest run, i think.

File: cmd/juju/caas/add.go (hunk 25 above)
Comment (ycliuhw) [reply cmd/juju/caas/add.go]: yeah, good finding, changed

File: cmd/juju/caas/add_test.go (hunk 26 above)
Comment (ycliuhw) [reply cmd/juju/caas/add_test.go]: ah, u r right. no need to have it in suite even.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**